### PR TITLE
Add drag-resizing with cursor feature

### PR DESCRIPTION
**Pull request by MrGibus**
- [X] Tested on all platforms changed
- [x] Compilation warnings were addressed
- [X] `cargo fmt` has been run on this branch
- [X] `cargo doc` builds successfully
- [X] Added an entry to `CHANGELOG.md` if knowledge of this change could be valuable to users
- [X] Updated documentation to reflect any user-facing changes, including notes of platform-specific behavior
- [X] Created or updated an example program if it would help users understand this functionality
- [X] Updated [feature matrix](https://github.com/rust-windowing/winit/blob/master/FEATURES.md), if new features were added or implemented

Fix for https://github.com/rust-windowing/winit/issues/725 on windows
Similar in scope to https://github.com/rust-windowing/winit/pull/1840

This has only been implemented for windows:  
This [comment](https://github.com/rust-windowing/winit/issues/725#issuecomment-873463281) offer insight as to how this would translate to Linux platforms.

As far as using this to make a borderless window in similar fashion to electron, the approach has several drawbacks:
- No window resizing from win-hotkeys
- No shadows on the borders

To solve those issues [this SO Question](https://stackoverflow.com/questions/39731497/create-window-without-titlebar-with-resizable-border-and-without-bogus-6px-whit) and [this windows dev tutorial](https://docs.microsoft.com/en-us/windows/win32/dwm/customframe?redirectedfrom=MSDN) seem to be the standard. I don't understand how this would fit into winit at this time however.

I briefly experimented with removing the title bar and not the border through winapi, which also seems to mostly solve the issues above and this may be preferable to hiding *all* decorations and implementing standard window functionality manually.

### CODE REVIEW

**Review by maroider — COMMENTED** [review]
Thanks a bunch for taking the time to implement this :D

I'd like to see an addition to the feature matrix for this feature.

> As far as using this to make a borderless window in similar fashion to electron, the approach has several drawbacks:
> 
>     * No window resizing from win-hotkeys
> 
>     * No shadows on the borders

Indeed. But that's something that has to be solved in some other PR(s).
In fact, the first problem is being solved in #1891, although progress has unfortunately stalled a bit.

File: src/platform_impl/windows/window_state.rs [hunk 1]
```diff
@@ -188,16 +188,22 @@ impl WindowFlags {
...
         if self.contains(WindowFlags::DECORATIONS) {
-            style |= WS_CAPTION | WS_MINIMIZEBOX | WS_BORDER;
-            style_ex = WS_EX_WINDOWEDGE;
+            style |= WS_CAPTION | WS_SYSMENU | WS_MINIMIZEBOX; // Note WS_BORDER IS INCLUDED IN WS_CAPTION
+            style_ex |= WS_EX_WINDOWEDGE;
+
+            // Nest inside Decorations to preserve borderless windows
+            if self.contains(WindowFlags::RESIZABLE) {
+                style |= WS_SIZEBOX | WS_MAXIMIZEBOX;
+            }
```
[the remark this answers is not in the document]
Comment (MrGibus) [reply src/platform_impl/windows/window_state.rs:206]: The main thing here is that the resizable bit is set to 1,

I had a brief look at https://github.com/rust-windowing/winit/pull/1891 and I think that by removing ```const NO_DECORATIONS_AND_MASK = !WindowFlags::RESIZABLE.bits;``` It's correcting the issue whereby resizeable is linked to the decorations. Which is the reason for my changes. 

Besides that the only other point of note is that WS_BORDER isn't required where WS_CAPTION is set as [The MS Docs](https://docs.microsoft.com/en-us/windows/win32/winmsg/window-styles ) state it's included. And similarly the WS_SYSMENU  is only available with WS_CAPTION. This refactoring I did shouldn't be needed for this functionality, but may lead to weird issues further down the road.

File: src/platform_impl/windows/window.rs [hunk 2]
```diff
@@ -380,6 +380,41 @@ impl Window {
...
+                y: points.y as SHORT,
+            };
+            winuser::ReleaseCapture();
+            winuser::PostMessageW(
+                self.window.0,
+                winuser::WM_NCLBUTTONDOWN, // Non-Client Left Button Down
+                ht_dir as WPARAM,
+                &points as *const _ as LPARAM,
+            );
+        }
```
[the remark this answers is not in the document]
Comment (MrGibus) [reply src/platform_impl/windows/window.rs]: Agreed, and done.

**Comment by MrGibus** [comment]
A few other PRs are failing the nightly wasm32 target build check. 
```usage of `--profile` requires `-Z unstable-options` ```
looks like a workflow bug?

**Comment by acid1103** [comment]
Hi all. It's been several months since this issue was last touched. I'm curious if this is something that's still being considered. I'm working on an app and recently added a custom (hacky) implementation of this feature for it, but something officially supported out-of-the-box would be great.

**Comment by MrGibus** [comment]
@acid1103 I ended up making my own borderless window proof of concept by using the windows API directly as the way winnit deals with the messaging system in the api doesn't lend itself well to rebuilding the decorations.

File: src/platform_impl/windows/window_state.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (amrbashir) [reply src/platform_impl/windows/window_state.rs:206]: https://github.com/rust-windowing/winit/pull/2419 has been merged and this PR should be unblocked now.

**Comment by MrGibus** [comment]
I think the best way of doing borderless windows on windows is to use the following on a `WM_CREATE` event:
(Using windows-rs)
```
/// with shadows
const BORDERLESS_MARGINS: MARGINS = MARGINS {
    cxLeftWidth: 1,
    cxRightWidth: 1,
    cyTopHeight: 1,
    cyBottomHeight: 1,
};

/// no shadows
const FLAT_MARGINS: MARGINS = MARGINS {
    cxLeftWidth: 0,
    cxRightWidth: 0,
    cyTopHeight: 0,
    cyBottomHeight: 0,
};

/// Primarily used to create borderless windows
/// https://microsoft.github.io/windows-docs-rs/doc/windows/Win32/Graphics/Dwm/fn.DwmExtendFrameIntoClientArea.html
/// https://docs.microsoft.com/en-us/windows/win32/api/dwmapi/nf-dwmapi-dwmextendframeintoclientarea
pub fn extend_frame_into_client_area(handle: HWND, style: &WindowStyle) -> Result<()> {
    let margins = match style {
        WindowStyle::FlatBorderless => FLAT_MARGINS,
        WindowStyle::Borderless => BORDERLESS_MARGINS,
    };

    if handle.is_invalid() {
        panic!("Cannot extend frame into client area due to invalid window handle")
    }

    unsafe { DwmExtendFrameIntoClientArea(handle, &margins) }
}
```
See below example using wgpu to render over borders:
https://github.com/MrGibus/windows_borderless_rust

**Comment by amrbashir** [comment]
Btw, this PR is superseded by https://github.com/rust-windowing/winit/pull/2515 and https://github.com/rust-windowing/winit/pull/2966

**Comment by daxpedda** [comment]
Thanks @amrbashir!

**Pull request closed without merging**